plutus-core/untyped-plutus-core/Language/UntypedPlutusCore/Evaluation/HOAS.hs

```diff
@@ -0,0 +1,307 @@
+    _InternalHoasError = _InternalEvaluationError . _InternalHoasError
+instance AsUnliftingError (InternalHoasError fun) where
+    _UnliftingError = _UnliftingHoasError
+
+-- | Convert an 'HTerm' into a 'Term' running all internal monadic actions along the way.
+fromHTerm :: Monad m => HTerm m name uni fun ann -> m (Term name uni fun ann)
+fromHTerm (HConstant ann val)     = pure $ Constant ann val
+fromHTerm (HBuiltin ann fun)      = pure $ Builtin ann fun
+fromHTerm (HVar ann name)         = pure $ Var ann name
+-- Here we do not recover the original annotation and instead use the one that the whole lambda
+-- is annotated with. We could probably handle annotations better, but we don't care for now.
+fromHTerm (HLamAbs ann name body) = LamAbs ann name <$> (body (HVar ann name) >>= fromHTerm)
+fromHTerm (HApply ann fun arg)    = Apply ann <$> fromHTerm fun <*> fromHTerm arg
+fromHTerm (HDelay ann getBody)    = Delay ann <$> (getBody () >>= fromHTerm)
+fromHTerm (HForce ann term)       = Force ann <$> fromHTerm term
+fromHTerm (HError ann)            = pure $ Error ann
+
+-- | Convert a 'Value' into a 'Term' running all internal monadic actions along the way and
+-- extracting all partial builtin applications.
+fromValue
+    :: Value unique name uni fun ann
+    -> EvalM unique name uni fun ann (Term name uni fun ann)
+fromValue = fromHTerm >=> bindFunM (const _builtinAppTerm)
+
+-- | Run an 'EvalM' computation and convert the cause of a possible error from 'Value' to 'Term'.
+runEvalM
+    :: EvalM unique name uni fun ann a
+    -> Either (HoasException fun (Term name uni fun ann)) a
+runEvalM = bimap errorValueToTerm id . unEvalM where
+    -- Here we call 'runEvalM' recursively. It's fine when the underlying monad is 'Either',
+    -- but if it had 'ReaderT', then we'd also need to make sure that 'runEvalM' is supplied
+    -- with the most recent environment, not the initial one.
+    errorValueToTerm = either id id . traverse (runEvalM . fromValue)
+
+lookupVar
+    :: HasUnique name unique
+    => ann -> name -> UniqueMap unique value -> EvalM unique name uni fun ann value
+lookupVar ann name env =
+    case lookupName name env of
+        Just term -> pure term
+        Nothing   ->
+            throwingWithCause _InternalHoasError FreeVariableHoasError $
+                Just $ HVar ann name
+
+-- | Retrieve the meaning of a built-in function.
+lookupBuiltin
+    :: (value ~ Value unique name uni fun ann, Ix fun)
+    => ann -> fun -> BuiltinsRuntime fun value -> EvalM unique name uni fun ann value
+lookupBuiltin ann fun (BuiltinsRuntime meanings) =
+    case meanings ^? ix fun of
+        Nothing      -> throwingWithCause _InternalHoasError (UnknownBuiltinHoasError fun) Nothing
+        Just meaning -> pure . HBuiltin ann $ BuiltinApp (pure $ Builtin ann fun) meaning
+{- Note [Handling non-constant results]
+Evaluation may return a non-constant term. This has a couple of implications:
```

Comment:
Non-value term? I'm not sure I actually see why this has to be the case. Yes, we can return a partially-applied builtin, but those are values, right? I guess this reduces to the question of why our `Value` type is the whole of `HTerm`...

Reply:
No, non-constant term, as in "not a `Constant`". Changed the wording to

> Evaluation may return a non-'Constant' term.

Reply:
Yes, I understood that. I meant to question whether the unusual thing was that the thing being returned was not a value - why would we expect it to be a constant?

Reply:
Yes, there's nothing unusual about the fact that we return general values and not just constants. The note is meant to clarify what consequences that has. Should I reword something?

Reply:
I guess it's just that "Handling non-constant results" makes me think that there's something special or unusual about that. But actually it's perfectly clear if you read it.